aws-iot-billinggroup/src/main/java/software/amazon/iot/billinggroup/CreateHandler.java

```diff
@@ -32,43 +41,63 @@ protected ProgressEvent<ResourceModel, CallbackContext> handleRequest(
+                                .progress())
+                .then(progress -> ProgressEvent.defaultSuccessHandler(resourceModel));
+    }
 
-        try {
-            try {
-                // check whether the resource exists - ResourceNotFound is thrown otherwise.
-                proxyClient.injectCredentialsAndInvokeV2(
-                        describeBillingGroupRequest,
-                        proxyClient.client()::describeBillingGroup
-                );
-                logger.log(String.format("%s %s already exists. Failing create handler.",
-                        ResourceModel.TYPE_NAME,describeBillingGroupRequest.billingGroupName()));
-                throw new ResourceAlreadyExistsException(new Exception(
-                        String.format("%s %s already exists. Failing create handler.",
-                        ResourceModel.TYPE_NAME,describeBillingGroupRequest.billingGroupName())));
-            } catch (ResourceNotFoundException resourceNotFoundException) {
-                logger.log(String.format("%s %s does not exist. Continuing to create resource.",
-                        ResourceModel.TYPE_NAME,describeBillingGroupRequest.billingGroupName()));
-            }
+    /**
+     * Implement client invocation of the create request through the proxyClient, which is already initialised with
+     * caller credentials, correct region and retry settings
+     *
+     * @param createBillingGroupRequest the aws service request to create a resource
+     * @param proxyClient the aws service client to make the call
+     * @return awsResponse create resource response
+     */
+    private CreateBillingGroupResponse createResource(
+            final CreateBillingGroupRequest createBillingGroupRequest,
+            final ProxyClient<IotClient> proxyClient) {
 
-            proxyClient.injectCredentialsAndInvokeV2(
+        try {
+            checkForBillingGroup(createBillingGroupRequest.billingGroupName(), proxyClient);
+            final CreateBillingGroupResponse createBillingGroupResponse = proxyClient.injectCredentialsAndInvokeV2(
                     createBillingGroupRequest, proxyClient.client()::createBillingGroup);
-            logger.log(String.format("%s %s successfully created.",
+            logger.log(String.format("%s [%s] successfully created.",
                     ResourceModel.TYPE_NAME, createBillingGroupRequest.billingGroupName()));
-        } catch (Exception e) {
-            return Translator.translateExceptionToProgressEvent(resourceModel, e, logger);
+            return createBillingGroupResponse;
+        } catch (IotException e) {
+            throw Translator.translateIotExceptionToHandlerException(createBillingGroupRequest.billingGroupName(), OPERATION, e);
         }
+    }
 
-        return ProgressEvent.defaultSuccessHandler(resourceModel);
+    private void checkForBillingGroup(final String billingGroupName, final ProxyClient<IotClient> proxyClient) {
```

Comment:
Can we move this method to BaseHandler? cause seems like its defined and called by both Delete and Create Handlers?

Reply:
Both the implementations of the method have different exception handling though. In create handler - it is meant to throw `CfnAlreadyExistsException` ~ signifying that that resource already exists and therefore CREATE operation cannot proceed. In delete handler - it is meant only to check the presence of the resource to verify that there is no drift in CFN template and resource status.

Reply:
oo i see, alright then